### DISCUSSION TITLE
Trouble with exclude in .chezmoiignore (likely misuse)

### DISCUSSION THREAD

**Discussion by jibbers42** [twpayne/chezmoi#1977] [Q&A] [answered]
My `.chezmoiignore` looks like...
```
{{ if ne .chezmoi.os "windows" -}}
Documents/
{{ end }}

{{ if ne .chezmoi.os "linux" -}}
.config/
{{- end }}

README.md

!.config/chezmoi/chezmoi.toml
!.config/powershell/completions/** # I've tried single * and no *
```

where I want the chezmoi.toml (and the powershell completions) to be in both windows and linux. 

for windows:
```
> chezmoi.exe managed
Documents
Documents/PowerShell
Documents/PowerShell/Microsoft.PowerShell_profile.ps1
```
(the .config/chezmoi/chezmoi.toml nor the completions are managed)

Any ideas on how to get this to work?

**Comment by jibbers42** [comment, marked as answer]
After looking at the docs again, I realize I wasn't doing chezmoi config quite right. Now that I have `.chezmoi.toml.tmpl` in the root of my source, instead of managing it as a normal config file, it's working again.

I'm still unsure how to exclude files from an ignored folder in `.chezmoiignore`, specifically the powershell completions in the original post.

**Reply by twpayne** [reply]
Generally, you need two lines in your `.chezmoiignore` file to ignore a directory and its contents, for example:

```
Documents     # to ignore the Documents directory
Documents/**  # to ignore everything in the Documents directory
```

This is a bit different to `.gitignore` files, as chezmoi needs to allow ignoring a directory entry but not ignoring some of the directory's contents.

Hope this helps. Please reply if this doesn't solve it for you.

**Reply by jibbers42** [reply]
Thanks for the note... hmm, the example given at https://www.chezmoi.io/reference/special-files-and-directories/chezmoiignore/ shows
```
backups/   # ignore backups folder in chezmoi directory and all its contents
backups/** # ignore all contents of backups folder in chezmoi directory
           # but not backups folder itself
```
suggesting that adding the slash ignores both. I assume that means this section of the docs is out of date?

My actual issue is understanding exclusions though. I'm not sure if I'm misunderstanding the feature or misunderstanding the usage. My expectation is if I ignore a folder and it's contents I can then exclude files in those contents to have them managed, for example (using your style above)...

.chezmoiIgnore
```
{{ if ne .chezmoi.os "linux" -}}
.config
.config/**
{{- end }}

# not sure how to write this, but even an exclude for one specific file doesn't work
!.config/powershell/completions
!.config/powershell/completions/**
!.config/powershell/completions/chezmoi.ps1
```

When I test on windows I do not get the completion files.

EDIT: fixed some commented out lines that I had while testing.

**Reply by JordanP-Dev** [reply]
I was also trying to figure out how to use exclusions to only allow whitelisted files for a host.  It looks like you need exclusions for all parent directories.  So, if you want to include everything in the powershell completions folder, I think the following would work:

```
!.config
!.config/powershell
!.config/powershell/completions
!.config/powershell/completions/* # Or ** if there are subdirectories you want to include
```

**Reply by ptere** [reply]
> Generally, you need two lines in your `.chezmoiignore` file to ignore a directory and its contents

As @jibbers42 pointed out, this [contradicts the documentation](https://github.com/twpayne/chezmoi/blob/46e13edd60e24af35d126ea4aa54c79b58877d79/assets/chezmoi.io/docs/reference/special-files-and-directories/chezmoiignore.md?plain=1#L30-L32). It would be great to get this resolved.

**Reply by twpayne** [reply]
Thanks for the reminder. Fixed with 5b738bde2e7d80969794ff1bea30a89fac091bca.

**Reply by Avishayy** [reply]
@twpayne may I ask why there's a differentiation between a folder and its contents? I'm not sure I understand, I'm a new chezmoi user

**Reply by twpayne** [reply]
Yes, sometimes you just want to manage a folder, for example to ensure that a folder is created and has the right permissions, and you don't care about its contents.

**Reply by Jimmy2027** [reply]
Is it possible to include only one file in a subfolder without including all files in that folder?
Let's say I want to exclude everything from my work machine except `.local/bin/my_bin`
I tried the following config:
```
{{- if or (eq .chezmoi.hostname "my_work_machine") }}
# don't manage these files on work machines
*
!.local/bin/my_bin
{{- end }}

``` 
but then `chezmoi managed` is empty
if I add the parent folders as described above, all files in those folders are added.

**Reply by twpayne** [reply]
The `*` in your `.chezmoiignore` applies to all directories.

Please try:

```
{{- if or (eq .chezmoi.hostname "my_work_machine") }}
# don't manage these files on work machines
.local/bin/**
!.local/bin/my_bin
{{- end }}
```

**Comment by needleshaped** [comment]
I had very similar requirement:

Ignore Documents, except for 
- Documents/PowerShell/Microsoft.PowerShell_profile.ps1 and 
- Documents/WindowsPowerShell/Microsoft.PowerShell_profile.ps1

Following worked out in `.chezmoiignore` ✅:
```
{{- if eq .chezmoi.os "windows" }}
Documents/*
!Documents/*PowerShell/
{{- end }}
```

- note `Documents/*`, not `**`. Double star did NOT work for me, 
- Neiter excluding specific file `!Documents/**PowerShell/Microsoft.PowerShell_profile.ps1` did NOT work, nor `!Documents/**PowerShell/*.ps1` or `!Documents/*PowerShell/*.ps1`
- Documentation https://www.chezmoi.io/reference/special-files/chezmoiignore/ need an update, it does not answer this quite simple use case. Added https://github.com/twpayne/chezmoi/pull/4635 providing example, but still, whole scenario is not clear, only trial & error helped.
